docs/testing.md

```diff
@@ -0,0 +1,98 @@
+Testing Strategies
+==================
+
+General Concepts
+------------------
+
+* security vs. reliability vs. performance tests
+* black-box vs. white-box
+
+Further Reading
+* [CMU: Software Testing](https://users.ece.cmu.edu/~koopman/des_s99/sw_testing/)
+* [OWASP: Security Testing Web Applications throughout Automated Software
+Tests](https://www.owasp.org/images/9/99/AutomatedSecurityTestingofWebApplications-StephendeVries.pdf)
+
+Validate Acceptance Criteria
+----------------------------
+
+The acceptance criteria formalize success or failure of user stories. To
+prevent incorrect implementation, also consider negative user stories (also
+called *evil user stories* or *abuser story*). The goal is to identify
+behavior (of the user as well as of the application) that is not intended
+initially. If not identified, it is unlikely, that any developer will
+write tests that consider the corresponding behavior.
+
+### Summary
+* identify unintended application behavior
+
+Further Reading
+* [OWASP: Agile Software Development: Don't Forget EVIL User Stories](https://www.owasp.org/index.php/Agile_Software_Development:_Don%27t_Forget_EVIL_User_Stories)
+* [DZone: Adding Appsec to Agile](https://dzone.com/articles/adding-appsec-agile-security)
+
+
+Coverage
+--------
+
+The code coverage measures how many lines of code are executed during
+testing. It's a good idea to maximize coverage. It's not a good idea
+to rely on the coverage. The coverage does not give any advice about
+logical issues or cases, that are not considered in your code. However,
+there should be a good reason for a low coverage. Having no time is
+not a good reason.
```

Comment:
I don't agree with the last sentence. Having no time even may be the only reason for low coverage. Mostly anything can be done with enough time, including refactoring the whole application in order to improve testability.

As an example: Our experience with browser-based integration tests in adhocracy3 was mostly that they were extremely unreliable. We could have spend a lot of time and effort to improve that, but other things were more impotant.